// @nestjs-addons/in-memory-db/src/v1/in-memory-db.module.js
"use strict";
var InMemoryDBV1Module_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryDBV1Module = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const services_1 = require("./services");
const providers_1 = require("./providers");
let InMemoryDBV1Module = InMemoryDBV1Module_1 = class InMemoryDBV1Module {
    static forRoot(config = {}) {
        const providers = providers_1.createInMemoryDBV1ForRootProviders(config);
        return {
            module: InMemoryDBV1Module_1,
            providers,
            exports: providers,
        };
    }
    static forFeature(featureName, config = {}) {
        const providers = providers_1.createInMemoryDBV1ForFeatureProviders(featureName, config);
        return {
            module: InMemoryDBV1Module_1,
            providers,
            exports: providers,
        };
    }
};
InMemoryDBV1Module = InMemoryDBV1Module_1 = tslib_1.__decorate([
    common_1.Module({
        providers: [services_1.InMemoryDBV1Service],
        exports: [services_1.InMemoryDBV1Service],
    })
], InMemoryDBV1Module);
exports.InMemoryDBV1Module = InMemoryDBV1Module;
