// @nestjs-addons/in-memory-db/src/controllers/in-memory-db.controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryDBEntityController = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
class InMemoryDBEntityController {
    constructor(dbService) {
        this.dbService = dbService;
    }
    create(record) {
        if (Array.isArray(record)) {
            return this.dbService.createMany(record);
        }
        return this.dbService.create(record);
    }
    update(id, record) {
        return this.dbService.update(Object.assign({ id }, record));
    }
    updateMany(records) {
        return this.dbService.updateMany(records);
    }
    delete(id) {
        return this.dbService.delete(id);
    }
    deleteMany(ids) {
        return this.dbService.deleteMany(ids);
    }
    get(id) {
        return this.dbService.get(id);
    }
    getMany(ids) {
        if (ids && Array.isArray(ids)) {
            return this.dbService.getMany(ids);
        }
        return this.dbService.getAll();
    }
}
tslib_1.__decorate([
    common_1.Post(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", Object)
], InMemoryDBEntityController.prototype, "create", null);
tslib_1.__decorate([
    common_1.Put(':id'),
    tslib_1.__param(0, common_1.Param('id')), tslib_1.__param(1, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", void 0)
], InMemoryDBEntityController.prototype, "update", null);
tslib_1.__decorate([
    common_1.Put(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Array]),
    tslib_1.__metadata("design:returntype", void 0)
], InMemoryDBEntityController.prototype, "updateMany", null);
tslib_1.__decorate([
    common_1.Delete(':id'),
    tslib_1.__param(0, common_1.Param('id')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], InMemoryDBEntityController.prototype, "delete", null);
tslib_1.__decorate([
    common_1.Delete(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Array]),
    tslib_1.__metadata("design:returntype", void 0)
], InMemoryDBEntityController.prototype, "deleteMany", null);
tslib_1.__decorate([
    common_1.Get(':id'),
    tslib_1.__param(0, common_1.Param('id')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", Object)
], InMemoryDBEntityController.prototype, "get", null);
tslib_1.__decorate([
    common_1.Get(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Array]),
    tslib_1.__metadata("design:returntype", Array)
], InMemoryDBEntityController.prototype, "getMany", null);
exports.InMemoryDBEntityController = InMemoryDBEntityController;
