// @nestjs-addons/in-memory-db/src/controllers/in-memory-db-async.controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryDBEntityAsyncController = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const rxjs_1 = require("rxjs");
class InMemoryDBEntityAsyncController {
    constructor(dbService) {
        this.dbService = dbService;
    }
    create(record) {
        if (Array.isArray(record)) {
            return this.dbService.createManyAsync(record);
        }
        return this.dbService.createAsync(record);
    }
    update(id, record) {
        return this.dbService.updateAsync(Object.assign({ id }, record));
    }
    updateMany(records) {
        return this.dbService.updateManyAsync(records);
    }
    delete(id) {
        return this.dbService.deleteAsync(id);
    }
    deleteMany(ids) {
        return this.dbService.deleteManyAsync(ids);
    }
    get(id) {
        return this.dbService.getAsync(id);
    }
    getMany(ids) {
        if (ids && Array.isArray(ids)) {
            return this.dbService.getManyAsync(ids);
        }
        return this.dbService.getAllAsync();
    }
}
tslib_1.__decorate([
    common_1.Post(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "create", null);
tslib_1.__decorate([
    common_1.Put(':id'),
    tslib_1.__param(0, common_1.Param('id')),
    tslib_1.__param(1, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "update", null);
tslib_1.__decorate([
    common_1.Put(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Array]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "updateMany", null);
tslib_1.__decorate([
    common_1.Delete(':id'),
    tslib_1.__param(0, common_1.Param('id')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "delete", null);
tslib_1.__decorate([
    common_1.Delete(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Array]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "deleteMany", null);
tslib_1.__decorate([
    common_1.Get(':id'),
    tslib_1.__param(0, common_1.Param('id')),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "get", null);
tslib_1.__decorate([
    common_1.Get(),
    tslib_1.__param(0, common_1.Body()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Array]),
    tslib_1.__metadata("design:returntype", rxjs_1.Observable)
], InMemoryDBEntityAsyncController.prototype, "getMany", null);
exports.InMemoryDBEntityAsyncController = InMemoryDBEntityAsyncController;
