// @nestjs-addons/in-memory-db/src/in-memory-db.module.js
"use strict";
var InMemoryDBModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryDBModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const services_1 = require("./services");
const providers_1 = require("./providers");
let InMemoryDBModule = InMemoryDBModule_1 = class InMemoryDBModule {
    static forRoot(config = {}) {
        const providers = providers_1.createInMemoryDBForRootProviders(config);
        return {
            module: InMemoryDBModule_1,
            providers,
            exports: providers,
        };
    }
    static forFeature(featureName, config = {}) {
        const providers = providers_1.createInMemoryDBForFeatureProviders(featureName, config);
        return {
            module: InMemoryDBModule_1,
            providers,
            exports: providers,
        };
    }
};
InMemoryDBModule = InMemoryDBModule_1 = tslib_1.__decorate([
    common_1.Module({
        providers: [services_1.InMemoryDBService],
        exports: [services_1.InMemoryDBService],
    })
], InMemoryDBModule);
exports.InMemoryDBModule = InMemoryDBModule;
