// @nestjs-addons/in-memory-db/src/services/in-memory-db.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryDBService = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const rxjs_1 = require("rxjs");
const uuid_1 = require("uuid");
let InMemoryDBService = class InMemoryDBService {
    constructor(config) {
        this.config = config;
        this.recordMap = {};
    }
    set records(records) {
        if (!records || records.length === 0) {
            this.recordMap = {};
        }
        this.recordMap = records.reduce((previous, current) => {
            return Object.assign(Object.assign({}, previous), { [current.id]: current });
        }, this.recordMap);
    }
    get records() {
        return Object.keys(this.recordMap).map((key) => this.recordMap[key]);
    }
    create(record, getNextId = () => uuid_1.v4()) {
        const id = record.id || getNextId();
        const newRecord = Object.assign(Object.assign({}, record), { id });
        this.recordMap = Object.assign(Object.assign({}, this.recordMap), { [newRecord.id]: newRecord });
        return newRecord;
    }
    createAsync(record) {
        const result$ = rxjs_1.of(this.create(record));
        return result$;
    }
    createMany(records, getNextId = () => uuid_1.v4()) {
        return records.map((record) => this.create(record, getNextId));
    }
    createManyAsync(records, getNextId = () => uuid_1.v4()) {
        const result$ = rxjs_1.of(this.createMany(records, getNextId));
        return result$;
    }
    update(record) {
        this.recordMap = Object.assign(Object.assign({}, this.recordMap), { [record.id]: Object.assign({}, record) });
    }
    updateAsync(record) {
        this.update(record);
        const result$ = rxjs_1.of();
        return result$;
    }
    updateMany(records) {
        for (const record of records) {
            this.update(record);
        }
    }
    updateManyAsync(records) {
        this.updateMany(records);
        const result$ = rxjs_1.of();
        return result$;
    }
    delete(id) {
        const _a = this.recordMap, _b = id, removed = _a[_b], remainder = tslib_1.__rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
        this.recordMap = Object.assign({}, remainder);
    }
    deleteAsync(id) {
        this.delete(id);
        const result$ = rxjs_1.of();
        return result$;
    }
    deleteMany(ids) {
        for (const id of ids) {
            this.delete(id);
        }
    }
    deleteManyAsync(ids) {
        this.deleteMany(ids);
        const result$ = rxjs_1.of();
        return result$;
    }
    get(id) {
        return this.recordMap[id];
    }
    getAsync(id) {
        const result$ = rxjs_1.of(this.get(id));
        return result$;
    }
    getMany(ids) {
        const records = ids
            .filter((id) => this.recordMap[id])
            .map((id) => {
            return this.recordMap[id];
        });
        return records;
    }
    getManyAsync(ids) {
        const result$ = rxjs_1.of(this.getMany(ids));
        return result$;
    }
    getAll() {
        return this.records || [];
    }
    getAllAsync() {
        const result$ = rxjs_1.of(this.getAll());
        return result$;
    }
    query(predicate) {
        return this.records.filter(predicate);
    }
    queryAsync(predicate) {
        const result$ = rxjs_1.of(this.query(predicate));
        return result$;
    }
    seed(recordFactory, amount = 10, getNextId = () => uuid_1.v4()) {
        amount = amount === null ? 10 : amount;
        const recordsToCreate = [...Array(amount).keys()].map((i) => recordFactory(i));
        this.createMany(recordsToCreate, getNextId);
    }
};
InMemoryDBService = tslib_1.__decorate([
    common_1.Injectable(),
    tslib_1.__param(0, common_1.Optional()),
    tslib_1.__metadata("design:paramtypes", [Object])
], InMemoryDBService);
exports.InMemoryDBService = InMemoryDBService;
